src/websockethandler.cpp

```diff
@@ -0,0 +1,178 @@
+ *
+ * No-op in case the connection is already closed.
+ *
+ * Closing the websocket connection through this function does keeps it closed,
+ * unless the connection is explicitly reopened through
+ * `WebSocketHandler::open`.
+ */
+void WebSocketHandler::close() {
+  if (m_webSocket.state() == QAbstractSocket::UnconnectedState ||
+      m_webSocket.state() == QAbstractSocket::ClosingState) {
+    qDebug() << "Attempted to close a WebSocket, but it's already closed.";
+    return;
+  }
+
+  qDebug() << "Closing WebSocket";
+  m_reconnectOnClose = false;
+
+  m_webSocket.close();
+}
+
+/**
+ * @brief Sends a ping to the WebSocket server.
+ *
+ * WebSocket connections may silently be broken. Periodically pinging the server
+ * keeps track of the connection. If verified that the connection has been
+ * broken the connection is properly closed, which triggers new connection
+ * attempts.
+ */
+void WebSocketHandler::ping() {
+  if (!this->m_isAlive) {
+    qDebug() << "WebSocket was disconnected silently. Closing.";
+    m_webSocket.close();
+    return;
+  }
+
+  m_isAlive = false;
+  m_webSocket.ping();
+}
+
+/**
+ * @brief Acknowledges a pong response from the WebSocket server.
+ *
+ * @param elapsedTime How long it took to get the response since a ping was sent
+ * out.
+ */
+void WebSocketHandler::onPong(quint64 elapsedTime) {
+  qDebug() << "WebSocket pong" << elapsedTime;
+  m_isAlive = true;
+}
+
+/**
+ * @brief Ackowledges the WebSocket has been succesfully connected.
+ */
+void WebSocketHandler::onConnected() {
+  qDebug() << "WebSocket connected";
+
+  m_isAlive = true;
+  connect(&m_webSocket, &QWebSocket::textMessageReceived, this,
+          &WebSocketHandler::onMessageReceived);
+
+  // Start ping timer.
+  m_pingTimer = new QTimer(this);
+  connect(m_pingTimer, &QTimer::timeout, this, &WebSocketHandler::ping);
+  m_pingTimer->start(WEBSOCKET_PING_INTERVAL);
+}
+
+/**
+ * @brief Ackowledges the WebSocket has been closed.
+ *
+ * May trigger a reconnection attempt depending on the `m_reconnectOnClose`
+ * state.
+ */
+void WebSocketHandler::onClose() {
+  m_pingTimer->stop();
+
+  if (m_reconnectOnClose) {
+    qDebug() << "WebSocket closed. Attempting to reopen in:"
+             << WEBSOCKET_RETRY_INTERVAL;
+
+    QTimer::singleShot(WEBSOCKET_RETRY_INTERVAL, this, &WebSocketHandler::open);
+  } else {
+    qDebug() << "WebSocket closed explicitly. Will not attempt to reopen.";
+  }
+}
+
+/**
+ * @brief Ackowledges there was a WebSocket error.
+ *
+ */
+void WebSocketHandler::onError(QAbstractSocket::SocketError error) {
+  qDebug() << "WebSocket error:" << error;
+}
+
+/**
+ * @brief Acknowledges the user state has changed.
+ *
+ * When the user has authenticated, a WebSocket connection is openened.
+ * Otherwise it's closed.
+ *
+ * The Guardian WebSocket does not accept unauthenticated users. Attempting to
+ * connect while unauthenticated would result is an infinite retry loop.
+ */
+void WebSocketHandler::userStateChanged() {
```

Comment:
Would you mind expanding on why? I am relying on the user state, because I need the auth token to connect to the websocket. The State enum has many more variants, UserState looks more straight forward for this case IMO.

Reply:
We discussed this offile and decided to keep using UserState for now. The reason for moving to StateMain was that it has a more loaded meaning. Not only does it mean the user is authenticated, it also means the VPN is ready for connection, subscription is valid and so on. 

Because we are not sure the scope of the messages that will be passed through this WebSocket connection, relying on UserState or State right now doesn't make too big a difference. State is more restrictive. If we ever see the need we can change to rely on it instead.